### DISCUSSION TITLE
Pnpm pnpm-lock.yaml with resolution tarball fied

### DISCUSSION THREAD

**Discussion by mrogelja** [aquasecurity/trivy#5903] [Bugs]
### Description

Pnpm lockfile parser produces a corrupted SBOM file with a package resolution field that contains a tarball url. 

### Desired Behavior

Producing a valid SBOM file. 

### Actual Behavior

SBOM file contains a package entry :

```json
{
  "name": "",
  "SPDXID": "SPDXRef-Package-694d938945c788b4",
  "supplier": "NOASSERTION",
  "downloadLocation": "git+/myapp",
  "licenseConcluded": "NONE",
  "licenseDeclared": "NONE",
  "copyrightText": "",
  "externalRefs": [
    {
      "referenceCategory": "PACKAGE-MANAGER",
      "referenceType": "purl",
      "referenceLocator": "pkg:npm/"
    }
  ],
  "attributionTexts": [
    "PkgID: @"
  ],
  "primaryPackagePurpose": "LIBRARY"
},
```

### Reproduction Steps

```bash
Try the repo command with the following pnpm-lock.yaml file :


-----------------------

lockfileVersion: '6.0'

dependencies:
  jquery:
    specifier: 3.6.0
    version: 3.6.0
  promise:
    specifier: 8.1.0
    version: 8.1.0

packages:

  /asap@2.0.6:
    resolution: {integrity: sha512-BSHWgDSAiKs50o2Re8ppvp3seVHXSRM44cdSsT9FfNEUUZLOGWVCsiWaRPWM1Znn+mqZ1OfVZ3z3DWEzSp7hRA==}
    dev: false

  /jquery@3.6.0:
    resolution: {integrity: sha512-JVzAR/AjBvVt2BmYhxRCSYysDsPcssdmTFnzyLEts9qNwmjmu4JTAMYubEfwVOSwpQ1I1sKKFcxhZCI2buerfw==, tarball: https://centralrepo:443/artifactory/api/npm/npm/tslib/-/tslib-2.4.0.tgz}
    dev: false

  /promise@8.1.0:
    resolution: {integrity: sha512-W04AqnILOL/sPRXziNicCjSNRruLAuIHEOVBazepu0545DDNGYHz7ar9ZgZ1fMU8/MA4mVxp5rkBWRi6OXIy3Q==}
    dependencies:
      asap: 2.0.6
    dev: false
```

### Target

Filesystem

### Scanner

None

### Output Format

SPDX

### Mode

None

### Debug Output

```bash
2024-01-09T12:40:20.812Z        DEBUG   ["cyclonedx" "spdx" "spdx-json" "github"] automatically enables '--list-all-pkgs'.
2024-01-09T12:40:20.812Z        DEBUG   Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2024-01-09T12:40:20.812Z        DEBUG   Ignore statuses {"statuses": null}
2024-01-09T12:40:20.812Z        INFO    "--format spdx" and "--format spdx-json" disable security scanning
2024-01-09T12:40:20.821Z        DEBUG   cache dir:  /root/.cache/trivy
2024-01-09T12:40:20.822Z        DEBUG   Walk the file tree rooted at '/myapp' in parallel
2024-01-09T12:40:20.823Z        DEBUG   Skipping directory: projects
2024-01-09T12:40:20.823Z        DEBUG   Skipping directory: storybook
2024-01-09T12:40:20.823Z        DEBUG   Skipping directory: typedoc
2024-01-09T12:40:21.127Z        DEBUG   OS is not detected.
```

### Operating System

RHEL 8

### Version

```bash
0.45.1
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @mrogelja 
Thanks for your report!

> /jquery@3.6.0:
    resolution: {integrity: sha512-JVzAR/AjBvVt2BmYhxRCSYysDsPcssdmTFnzyLEts9qNwmjmu4JTAMYubEfwVOSwpQ1I1sKKFcxhZCI2buerfw==, tarball: https://centralrepo:443/artifactory/api/npm/npm/tslib/-/tslib-2.4.0.tgz}

This is a strange case. Tarball name contains `tslib`, but the dependency name is `jquery`.
Is this a valid lock file?

Typically packages installed from archives use the `name` and `version` fields.
Can you send pnpm commands to reproduce this case?

**Reply by mrogelja** [reply]
Hello,

My bad, I copy-pasted a bit quickly a tarball URL I found in my private repo into your testing lock file available [here](https://github.com/nikpivkin/go-dep-parser/blob/main/pkg/nodejs/pnpm/testdata/pnpm-lock_v6.yaml).

But the idea remains the same. The parser seems to have a problem when a tarball reference is present.

**Reply by DmitriyLewen** [reply]
Can you write the reproduction steps or create test repository for exploration?
The last time I updated pnpm logic, I was playing with tarballs.
In all cases where I installed dependencies using tarball, pnpm used `name` and `version` fields.

**Comment by DmitriyLewen** [comment]
Hello @mrogelja 
Although it is not entirely clear how to reproduce your example - #6034 should solve your problem too.

Regards, Dmitriy
